ddtrace/internal/injection.py

```diff
@@ -0,0 +1,139 @@
+from types import FunctionType
+from typing import Any
+from typing import Callable
+from typing import List
+from typing import Tuple
+
+from bytecode import Bytecode
+from bytecode import Instr
+
+from ddtrace.internal.utils.importlib import func_name
+
+
+_HookType = Callable[[Any], Any]
+
+
+HOOK_ARG_PREFIX = "_hook_arg"
+
+
+class InvalidLine(Exception):
+    """
+    Raised when trying to inject a hook on an invalid line, e.g. a comment or a blank line.
+    """
+
+
+def _inject_hook(code, hook, lineno, arg):
+    # type: (Bytecode, _HookType, int, Any) -> None
+    for i, instr in enumerate(code):
+        try:
+            if instr.lineno == lineno:
+                # gotcha!
+                break
+        except AttributeError:
+            # pseudo-instruction (e.g. label)
+            pass
+    else:
+        raise InvalidLine("Line %d is blank or a comment line" % lineno)
+
+    # actual injection
+    code[i:i] = Bytecode(
+        [
+            Instr("LOAD_CONST", hook, lineno=lineno),
+            Instr("LOAD_CONST", arg, lineno=lineno),
+            Instr("CALL_FUNCTION", 1, lineno=lineno),
+            Instr("POP_TOP", lineno=lineno),
+        ]
+    )
+
+
+def _eject_hook(code, line, arg):
+    # type: (Bytecode, int, Any) -> None
+    for i, instr in enumerate(code):
+        try:
+            if (
+                instr.lineno == line
+                and code[i + 1].arg is arg
+                and [code[_].name for _ in range(i, i + 4)] == ["LOAD_CONST", "LOAD_CONST", "CALL_FUNCTION", "POP_TOP"]
+            ):
```

Comment:
a comment here seems useful, took me a second to realize that `arg` is the actual indicator that this is a hook and not just some coincidence that the instructions lines up.